### PR TITLE
Remove rustc_args_required_const attribute

### LINKED ISSUE #69493: Remove special-case for `simd_shuffle` intrinsics in promotion
#69280 tried to normalize promotion of arguments that must be const by requiring `#[rustc_args_required_const]` everywhere. Currently, intrinsics beginning with the string "simd_shuffle" automatically get their third argument promoted. However, this broke the latest release of the `packed_simd` crate, which does not define the intrinsics with the requisite attribute (see #69313).

Although this is allowed breakage (you can't define intrinsics without the nightly-only `platform_intrinsics` feature), I reverted #69280 for now, since it's just a small cleanup and isn't blocking anything. I opened rust-lang/packed_simd#278 to fix `packed_simd`. Once a version of that crate is released with the fix, we should reapply #69280.

### PR DESCRIPTION
**Pull request by RalfJung**
Now that stdarch no longer needs it (thanks @Amanieu!), we can kill the `rustc_args_required_const` attribute. This means that lifetime extension of references to temporaries is the only remaining job that promotion is performing. :-)

r? @oli-obk 
Fixes https://github.com/rust-lang/rust/issues/69493

### CODE REVIEW

**Comment by oli-obk** [comment]
Looks like we need to do the same transformation that was done to stdarch for SIMD shuffle?

**Comment by RalfJung** [comment]
I am confused... so `aho-corasick` is somehow directly calling these intrinsics?

The attribute was removed from the shuffle intrinsics with https://github.com/rust-lang/stdarch/pull/1113, but it looks like they were special-cased in the promotion code so they got the magic treatment even without the attribute.

**Comment by oli-obk** [comment]
codegen does have a special case for these, yea. So... we need a const item with the indices and use that as the argument? Then bump stdsimd again and then we can do this PR?

**Comment by RalfJung** [comment]
Not just codegen, but promotion has a special case where it treats them as `rustc_args_required_const` even when that attribute is not present.

**Comment by RalfJung** [comment]
But yeah, I think the shuffle functions in stdarch need to become `rustc_legacy_const_generics` as well.

**Comment by RalfJung** [comment]
FWIW the error points here:

https://github.com/rust-lang/stdarch/blob/c2f21d24b938483d74e277d17fed81a6e7296a97/crates/core_arch/src/x86/avx2.rs#L298

**Comment by RalfJung** [comment]
Hm okay, this is a bit more tricky than I thought. The shuffle function cannot literally use the "legacy" attribute:
```
error: #[rustc_legacy_const_generics] functions must only have const generics
```
And moreover most (all?) calls are inside the same crate so the attribute does not even help...

The alternative is to fix all call sites of `simd_shuffle` to pass actual constants. I guess ideally the type checker would enforce this as part of a special check for these intrinsics?

**Comment by RalfJung** [comment]
I started seeing what it takes to fix this, and... it's a lot.^^ [Here's my WIP patch](https://github.com/RalfJung/stdarch/commit/a2569aab06ea8d8284379c5e1e0704e0ce1857cb). I don't like using `inline_const`; I feel the easiest way forward here might be a macro that generates code like
```rust
            const IDX: [u32; 32] = $idx;
            simd_shuffle32($x, $y, IDX)
```
The shuffle functions are AFAIK internal to stdarch, so we can pretty much do whatever we want here.

**Comment by RalfJung** [comment]
Yeah the macro approach [looks pretty good](https://github.com/RalfJung/stdarch/commit/3692d394278281e361e2570e82f4357426857bb8), I think.

However, we'll need to figure out what to do with code [like this](https://github.com/rust-lang/stdarch/blob/35595690d7881b20bd419875c2b64f284534d237/crates/core_arch/src/x86/avx2.rs#L423) that computes the mask based on other const-generic arguments...

**Comment by RalfJung** [comment]
> My proposal would be to compute all four constants as separate constants and do a runtime match on the generic arg and have four monomorphic calls this way.

Which "four constants"?
I don't understand your proposal.^^

I think long-term the intrinsic itself should be changed to use const generics; this also ensures that we don't get these ICEs. Short-term we can model this with const generic wrapper like
```rust
pub fn simd_shuffle32<T, U, IDX: const [u32; 32]>(x: T, y: T) -> U { 
  extern "platform-intrinsic" {
    fn simd_shuffle32<T, U>(x: T, y: T, idx: [u32; 32]) -> U;
  }
  simd_shuffle32(x, y, IDX)
}
```
However, this complains that `[u32; 32]` can't be used as const generic yet (and I assume we'd need another `incomplete_feature` to get around that).

What I don't know is how to write functions [like this](https://github.com/rust-lang/stdarch/blob/35595690d7881b20bd419875c2b64f284534d237/crates/core_arch/src/x86/avx2.rs#L423) that call such a const-generic shuffle-wrapper.

**Comment by oli-obk** [comment]
> Which "four constants"? There are 256 possible values for IMM8, aren't there?
I don't understand your proposal.^^

Yes, but it is bitmasked to 0b11

> What I don't know is how to write functions like this that call such a const-generic shuffle-wrapper.

That's what I was describing in my previous post. If you want to skip the runtime part and do it at compile time, you need the `const_evaluatable_checked` feature.

**Comment by RalfJung** [comment]
With [this patch](https://github.com/RalfJung/stdarch/commit/a38077c7156e2c94cf203fe2ab6294aa9b003cea) to stdarch, the bootstrap build completes again (or at least, aho-corasick builds; the full build is still ongoing).

@oli-obk @Amanieu would that be an okay avenue to take here?

**Comment by RalfJung** [comment]
@oli-obk I'd also like to have a reasonable error instead of the ICE when a shuffle argument is not a constant.

Currently, such an error is emitted here:
https://github.com/rust-lang/rust/blob/1b30245ea1286df96d673015c4519c861e06977a/compiler/rustc_mir/src/transform/promote_consts.rs#L795

But with promotion not having anything to do here any more, that makes little sense. What would be the right place to do that check? I thought it might be https://github.com/rust-lang/rust/blob/003b8eadd7a476c51956fe447894532d6e21937e/compiler/rustc_typeck/src/check/intrinsic.rs, but that file only checks the *signature* used in the `extern` block; we need something that checks each use site.

**Comment by oli-obk** [comment]
> but that function only checks the _signature_ used in the `extern` block; we need something that checks each use site.

mir validation perhaps? or just codegen, which is already checking that i thought

**Comment by RalfJung** [comment]
Checking during codegen means we only check this when stdarch is used; I'd rather get the build errors in stdarch itself than waiting until some crate runs into the ICE.^^

MIR validation runs a lot, doesn't it? Seems excessive.
I was thinking of putting it into the lower-intrinsics pass, since at least that is already intrinsics-related...^^

**Comment by oli-obk** [comment]
> 
> 
> Checking during codegen means we only check this when stdarch is used; I'd rather get the build errors in stdarch itself than waiting until some crate runs into the ICE.^^
> 
> MIR validation runs a lot, doesn't it? Seems excessive.
> I was thinking of putting it into the lower-intrinsics pass, since at least that is already intrinsics-related...^^

Yea, that opt is always run, even without optimizations, so 👍

**Comment by RalfJung** [comment]
> Yea, that opt is always run, even without optimizations, so +1

All right, done.

The stdarch PR is on its way at https://github.com/rust-lang/stdarch/pull/1160, but some arch/neon intrinsics still need to be adjusted (they are auto-generated and the generating tool needs some changes).

File: src/test/ui/issues/issue-38074.rs [hunk 1]
```diff
@@ -1,20 +0,0 @@
...
-}
-
-#[repr(simd)]
-#[derive(Clone, Copy)]
-#[allow(non_camel_case_types)]
-struct u64x2(u64, u64);
-
-fn main() {
-    let a = u64x2(1, 2);
-    let r: u64x2 = unsafe { simd_shuffle2(a, a, [0-0, 0-0]) };
```
Comment (RalfJung) [inline src/test/ui/issues/issue-38074.rs:17]: I deleted this test since it was specifically testing promotion, which we no longer do here.

**Comment by RalfJung** [comment]
CI is green. :)

**Comment by RalfJung** [comment]
Needs https://github.com/rust-lang/stdarch/pull/1162

**Pull request merged**